karapace/backup/v3/schema.proto

```diff
@@ -0,0 +1,152 @@
+// The version 3 backup format of Karapace uses Protobuf serialization. This
+// data exchange format makes it possible to verify backward and forward
+// compatibility statically, and compared to a custom format enables us to add
+// new fields without introducing a new version.
+//
+// The backup file format itself is as follows:
+//
+// 1. The first byte is the “magic” marker, the NUL byte (`\x00`). This ensures
+//    that other tools reading the file know that this is a binary file (e.g.
+//    the `file` utility).[^1]
+// 2. The second byte is the backup format version, an unsigned 8-bit integer.
+//    This is always `\x03` for the backup format 3.[^1][^2]
+// 3. The next 4 bytes contain the byte count of the `Metadata` message that
+//    follows. The length is an unsigned 32-bit little-endian integer.[^3]
+// 4. The `Metadata` message.
+// 5. Zero or more records follow:
+//    1. 4 bytes containing the unsigned 32-bit little-endian integer byte count
+//       of the `Envelope` message that follows.[^3]
+//    2. The `Envelope` message.
```

Comment:
For now the magic marker is needed. But could the rest of the file have TLV-format instead of just `<length><record>`? I am proposing: `[<tag><length><record>]`. Here it would be `<metadata_tag><length><metadata record>[<envelope_tag><length><envelope record>]`

Reply:
I think we should introduce `v4` if we want to change the file layout itself. Adding tags means that we need to manage the back-/forward compatibility for them somehow, and we have no tooling for that, unless we write it. Conventions are good, but they break apart easily. Using a fixed format means that there is no leeway, and thus nothing to manage.

I don't see us ever adding anything in between. Adding more fields to any of the messages is already possible, thanks to Protobuf, and with buf we also have the tooling to verify back-/forward compatibility of all changes.

Unless you have other reasons for proposing tags?

Reply:
Offline discussion, tags not needed. Rely on the Protobuf to handle the back-/forward compatibility.